ethyca/fides
# Initialize prefetched TCF experience preferences from cookie

Closes https://github.com/ethyca/fides/issues/4046
Closes https://github.com/ethyca/fides/issues/4011

### Description Of Changes

Makes prefetch work for TCF—the thing that was making this not work was that we weren't saving TCF preferences to a cookie, and so we couldn't initialize a user's preferences from a cookie (since a prefetched experience won't have a user's individual preferences). 


### Code Changes

* [x] Adds function to help save TCF preferences to a cookie
* [x] Updates assigning cookie values to an experience object function handle the TCF parts
* [x] Re-enable prefetching with TCF in the privacy center endpoint (there was a protective if statement that made it not possible to do both prefetch and TCF)

### Steps to Confirm

* Add the following to a `.env` file at the root of your `fides` repo (I'm actually not sure that _all_ of these are needed, but this works!)
```sh
FIDES_SAMPLE_APP__PRIVACY_CENTER_URL="http://localhost:3001"
FIDES_PRIVACY_CENTER__DEBUG=true
FIDES_PRIVACY_CENTER__IS_OVERLAY_ENABLED=true
FIDES_PRIVACY_CENTER__IS_GEOLOCATION_ENABLED=true
FIDES_PRIVACY_CENTER__IS_PREFETCH_ENABLED=true
FIDES_PRIVACY_CENTER__TCF_ENABLED=true
FIDES_PRIVACY_CENTER__GEOLOCATION_API_URL="https://cdn-api.ethyca.com/location"
FIDES_PRIVACY_CENTER__PRIVACY_CENTER_URL="http://localhost:3001"
```
* Run `nox -s "fides_env(test)"
* Once everything comes up, visit `localhost:8080/docs` and turn TCF on via the PATCH to `consent-settings`. You'll need to authorize with [these creds ](https://github.com/ethyca/fides/blob/aking/prod-1063/tcf-prefetch/src/fides/data/sample_project/fides.toml/#L19-L20) first
![image](https://github.com/ethyca/fides/assets/24641006/af543f55-c779-4270-8f80-6d09cf05cdc3)
* Now that TCF is enabled, visit `localhost:8080` and edit one of the existing systems.
* Add a data use such as `analytics.reporting.campaign_insights` (which is one of the TCF data uses). Give it a legal basis of "Consent" and save.
* Now visit cookie house at `localhost:3000` and click on "Frankfurt" from the geolocation dropdown. 
* Open the network tab and refresh the page. You should _not_ see a call to `/privacy-experience` since it is being prefetched by the privacy center! But you should still see the banner pop up
* Click on "manage preferences" in the banner, then "manage preferences" again to move past the initial layer. Toggle the the purpose `Understand audiences through statistics or combinations of data from different sources` ON
* Refresh the page. The banner should not pop up this time since there is nothing new to agree to
* Open the modal by clicking "Manage preferences"  at the bottom and "manage preferences" again in the modal to get to the second layer
* Your preference should still be saved (opted in)
* You can also look at your saved cookie, which should look something like
```
{
  "consent": {},
  "identity": {
    "fides_user_device_id": "9a82a38d-f5d4-477e-92c0-79e4233e2a2e"
  },
  "fides_meta": {
    "version": "0.9.0",
    "createdAt": "2023-09-21T15:21:31.361Z",
    "updatedAt": "2023-09-21T15:21:45.363Z"
  },
  "tcfConsent": {
    "purpose_preferences": { "9": true },
    "special_purpose_preferences": {},
    "feature_preferences": {},
    "special_feature_preferences": {},
    "vendor_preferences": {},
    "system_preferences": { "ctl_af18ae4b-8c7d-47e7-b3db-a13e5c319662": true }
  },
  "tcString": "CPyc0sAPyc0sAAMABBENASEAAACAAAAAAAAAAAAAAAAA.IgoMV_H3_bX9v8Tq16ft0eYFX99j77uQxBhfJs-4FyDvW_JwVn2EyNAUytqYKmRIEuzZBIQFsHJnURVChQogVryHsYkGMhTEAJ-BkgHMRa2dYCFxvmYtjEQIZ5_p_czf52R_9_dn-3czjT9lnnzVtO2dlcKidS59tH_l_bRKb-7Ied-7W_4v09tfrkW_eTVu_9evvp9-u_s_v__934AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQ"
}

```


### Pre-Merge Checklist
